hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/common/AbstractStreamWriteFunction.java

```diff
@@ -168,6 +148,7 @@ public void snapshotState(FunctionSnapshotContext functionSnapshotContext) throw
     snapshotState();
     // Reload the snapshot state as the current state.
     reloadWriteMetaState();
+    this.checkpointId = functionSnapshotContext.getCheckpointId();
```

Comment:
if user set tolerable checkpoint failure，

Step0: TaskA, TaskB and TaskC writing data using instant0

Step1: Checkpoint ID 1 trigger ==> taskA not finished(hang) , TaskB and TaskC finish snapthostState,and get next instant2 based on ckp id1.

Step2: checkpoint expiration due to a timeout.

Step3: Checkpoint ID 2 trigger ==> taskA still hang,  TaskB and TaskC finish snapthostState, and get next instant3 based on ckp id2.

Step4: taskA finish snapshot and get next instant3 based on ckp id2.

So ckpid1 will never finish and instant2 will never commit which means related data are lost

Reply:
okay，I see related logic in 
```
  private boolean commitInstants(long checkpointId) {
    // use < instead of <= because the write metadata event sends the last known checkpoint id which is smaller than the current one.
    List<Boolean> result = this.eventBuffers.getEventBufferStream().filter(entry -> entry.getKey() < checkpointId)
        .map(entry -> commitInstant(entry.getKey(), entry.getValue().getLeft(), entry.getValue().getRight())).collect(Collectors.toList());
    return result.stream().anyMatch(i -> i);
  }
```
which means coordinator will commit all the instant together before `checkpointId`. 

Maybe the situation just mentioned doesn't exist, but please double check here.